pkg/gardenlet/controller/vpaevictionrequirements/reconciler.go

```diff
@@ -0,0 +1,164 @@
+	defer cancel()
+
+	vpa := &vpaautoscalingv1.VerticalPodAutoscaler{}
+	err := r.SeedClient.Get(seedCtx, request.NamespacedName, vpa)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	log.Info("Reconciling vpa", "namespace/name", request.NamespacedName)
+
+	// double check just for fun: does the vpa have our label?
+	if metav1.HasLabel(vpa.ObjectMeta, constants.LabelVPAEvictionRequirementDownscaleInMaintenanceOnly) {
+		log.Info("Found the label " + constants.LabelVPAEvictionRequirementDownscaleInMaintenanceOnly)
+		return r.reconcileVPAForDownscaleInMaintenanceOnly(seedCtx, vpa)
+	} else if metav1.HasLabel(vpa.ObjectMeta, constants.LabelVPAEvictionRequirementDownscaleDisabled) {
+		log.Info("Found the label " + constants.LabelVPAEvictionRequirementDownscaleDisabled)
+		return r.reconcileVPAForDownscaleDisabled(seedCtx, vpa)
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *Reconciler) reconcileVPAForDownscaleInMaintenanceOnly(seedCtx context.Context, vpa *vpaautoscalingv1.VerticalPodAutoscaler) (reconcile.Result, error) {
+	log := logf.FromContext(seedCtx)
+	if !metav1.HasAnnotation(vpa.ObjectMeta, constants.AnnotationShootMaintenanceWindow) {
+		err := errors.New("didn't find maintenance window annotation on VPA, but VPA had label to be downscaled in maintenance only")
+		return reconcile.Result{}, err
+	}
+
+	windowAnnotation := vpa.GetAnnotations()[constants.AnnotationShootMaintenanceWindow]
+	splitWindowAnnotation := strings.Split(windowAnnotation, ",")
+	maintenanceTimeWindow, err := timewindow.ParseMaintenanceTimeWindow(splitWindowAnnotation[0], splitWindowAnnotation[1])
+	if err != nil {
+		log.Error(err, "Error during parsing the maintenance window from annotation", "begin", splitWindowAnnotation[0], "end", splitWindowAnnotation[1])
+		return reconcile.Result{}, err
+	}
+	isNowInMaintenanceTimeWindow := maintenanceTimeWindow.Contains(r.Clock.Now())
+
+	if isNowInMaintenanceTimeWindow {
+		log.Info("Shoot is inside maintenance window, removing the EvictionRequirement to allow downscaling", "shoot-namespace", vpa.GetNamespace(), "maintenanceWindow", maintenanceTimeWindow)
+		_, err := controllerutils.GetAndCreateOrMergePatch(seedCtx, r.SeedClient, vpa, func() error {
+			if !hasUpscaleOnlyEvictionRequirement(vpa.Spec.UpdatePolicy.EvictionRequirements) {
+				return nil
+			}
+			vpa.Spec.UpdatePolicy.EvictionRequirements = []*vpaautoscalingv1.EvictionRequirement{}
+			return nil
+		})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// requeue when the maintenance window ends, such that we can add the EvictionRequirement again
+		endTime := maintenanceTimeWindow.AdjustedEnd(r.Clock.Now())
+		requeueAfter := endTime.Sub(r.Clock.Now())
+		log.Info("Requeueing VPA", "requeueAfter", requeueAfter)
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
```

Comment:
Not sure how strict we want to be about the maintenance window. This might lead to the `vpa` object only being adapted with eviction requirements somewhat after the maintenance windows. The result might be a change of the managed resource, i.e. pod eviction, outside of the maintenance window, even if it just shortly after.

Do we want to include a (time) buffer to ensure that the change of the `vpa` resource is executed within the maintenance window? If yes, the calculation above also needs to be adapted as well.

Reply:
I wouldn't make this more complicated for now, given that one reconciliation loop runs in fractions of a second. I'm happy to introduce some buffer logic if we think that is necessary though. Any strong opinions on that or should we leave it as-is for now?

Reply:
Like I said I am not sure how strict we are in general about the maintenance window. In general, the next reconcile may happen at a somewhat later point in time in case there is a lot of load. Therefore, it is hard to ensure "correct" behaviour.

@timuthy Do you know how we handled cases like these in the past? Do we go 110% in ensuring that changes are confined to the maintenance windows or is it rather a best effort?

Reply:
This has just been on best effort basis, Gardener subtracts 15 Minutes from the end time to consider an active maintenance time window ([ref](https://github.com/gardener/gardener/blob/318124aecea1fde0e882f6963633b18066add9f9/pkg/utils/gardener/shoot.go#L137-L143)).

Reply:
We don't have this in the current HVPA scenario, so I'm not sure if we need to do this here.

Reply:
Ok, then let's leave it as is for now.